### PR TITLE
Improve frontend handling of many projects

### PR DESCRIPTION
**Pull request by filiptronicek**
## Description

Make the experience of using a large amount of projects / repository configurations more delightful.

- Removes the project context previously used by feature flags on the old `/project` paths
- Gets rid of `useListAllProjectsQuery`, which was used in ways it shouldn't have (like getting the current project)

## Related Issue(s)

Fixes ENT-717

## How to test

First, [join my org](https://ft-improve53bed5c2a8.preview.gitpod-dev.com/orgs/join?inviteId=f3e9715a-0e52-4ffe-9370-008d9a52c3d1) so you have some repos imported

Try starting a workspace:
1. By selecting an SCM repo
2. By selecting an imported repo
3. By directly going to the page with an imported repo's URL 

- [x] /werft with-preview

### CODE REVIEW

File: components/dashboard/src/data/git-providers/suggested-repositories-query.ts [hunk 1]
```diff
@@ -24,6 +24,9 @@ export const useSuggestedRepositories = ({ excludeConfigurations }: Props) => {
             const { repositories } = await scmClient.listSuggestedRepositories({
                 organizationId: org.id,
                 excludeConfigurations,
+                pagination: {
+                    pageSize: 100,
```
Comment (filiptronicek) [inline components/dashboard/src/data/git-providers/suggested-repositories-query.ts:28]: We now query at most 100 projects (SCM repos not included) for the initial load in repo selectors. Pagination is not implemented, but also not necessary yet

File: components/dashboard/src/data/git-providers/unified-repositories-search-query.ts
Comment (filiptronicek) [inline components/dashboard/src/data/git-providers/unified-repositories-search-query.ts:1]: The unified repo search has been changed to:
- in addition to the SCM repo search, searching for any imported repository with the search term
- also in addition to the above, we search for the currently selected imported repo either by ID or URL to resolve anything you might already have in there

File: components/gitpod-db/src/typeorm/project-db-impl.ts [hunk 2]
```diff
@@ -100,10 +111,9 @@ export class ProjectDBImpl extends TransactionalDBImpl<ProjectDB> implements Pro
         if (normalizedSearchTerm) {
             queryBuilder.andWhere(
                 new Brackets((qb) => {
-                    qb.where("project.cloneUrl LIKE :searchTerm", { searchTerm: `%${normalizedSearchTerm}%` }).orWhere(
-                        "project.name LIKE :searchTerm",
-                        { searchTerm: `%${normalizedSearchTerm}%` },
-                    );
+                    qb.where("project.cloneUrl LIKE :searchTerm", { searchTerm: `%${normalizedSearchTerm}%` })
+                        .orWhere("project.name LIKE :searchTerm", { searchTerm: `%${normalizedSearchTerm}%` })
+                        .orWhere("project.id LIKE :searchTerm", { searchTerm: `%${normalizedSearchTerm}%` });
```
Comment (filiptronicek) [inline components/gitpod-db/src/typeorm/project-db-impl.ts:116]: Get rid of this

File: components/dashboard/src/data/git-providers/unified-repositories-search-query.ts [hunk 3]
```diff
@@ -18,17 +27,55 @@ type UnifiedRepositorySearchArgs = {
...
     excludeConfigurations = false,
     onlyConfigurations = false,
     showExamples = false,
+    selectedContextURL,
+    selectedConfigurationId,
 }: UnifiedRepositorySearchArgs) => {
     const suggestedQuery = useSuggestedRepositories({ excludeConfigurations });
     const searchLimit = 30;
     const searchQuery = useSearchRepositories({ searchString, limit: searchLimit });
+    const configurationSearch = useListConfigurations({
```
Comment (filiptronicek) [inline components/dashboard/src/data/git-providers/unified-repositories-search-query.ts:45]: add comments here

File: components/dashboard/package.json [hunk 4]
```diff
@@ -97,7 +97,7 @@
     "react-scripts": "^5.0.1",
     "tailwind-underline-utils": "^1.1.2",
     "tailwindcss-filters": "^3.0.0",
-    "typescript": "^5.4.5",
+    "typescript": "^5.5.4",
```
Comment (filiptronicek) [inline components/dashboard/package.json:100]: doing this because I was using a feature introduced in 5.5

Comment (mustard-mh) [reply components/dashboard/package.json:100]: Just curious, which feature?

**Comment by mustard-mh** [comment]
Workspaces worked well
- By selecting an SCM repo https://mustardmh-springpetclin-0xi06n0lptu.ws-dev.ft-improve53bed5c2a8.preview.gitpod-dev.com/
- By selecting an imported repo https://tes-tenk100-va8u05g3lsd.ws-dev.ft-improve53bed5c2a8.preview.gitpod-dev.com/
- By directly going to the page with an imported repo's URL https://tes-2krepos1000-vyclwfugewd.ws-dev.ft-improve53bed5c2a8.preview.gitpod-dev.com/

File: components/dashboard/package.json (hunk 4 above)
Comment (filiptronicek) [reply components/dashboard/package.json:100]: It's the [Inferred Type Predicates](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-5.html#inferred-type-predicates), which lets us do `filter` with `!== undefined` without having to use something like `repo is SuggestedRepository`, basically typescript being smart about the fact that when you check something isn't undefined, its type correctly changes.

File: components/gitpod-db/src/typeorm/project-db-impl.ts [hunk 5]
```diff
@@ -69,9 +69,20 @@ export class ProjectDBImpl extends TransactionalDBImpl<ProjectDB> implements Pro
...
         const repo = await this.getRepo();
-        return repo.find({ where: { teamId: orgId, markedDeleted: false }, order: { name: "ASC" } });
+
+        const queryBuilder = repo
+            .createQueryBuilder("project")
+            .where("project.teamId = :teamId", { teamId: orgId })
+            .orderBy("project.creationTime", "DESC")
+            .andWhere("project.markedDeleted = false");
+
+        if (limit) {
```
Comment (mustard-mh) [inline components/gitpod-db/src/typeorm/project-db-impl.ts:81]: nit (not-blocking: Do you think we need a default value and max value (where we use it)?

File: components/dashboard/src/data/git-providers/unified-repositories-search-query.ts [hunk 6]
```diff
@@ -26,9 +35,34 @@ export const useUnifiedRepositorySearch = ({
     onlyConfigurations = false,
     showExamples = false,
 }: UnifiedRepositorySearchArgs) => {
-    const suggestedQuery = useSuggestedRepositories({ excludeConfigurations });
+    // 1st data source: suggested SCM repos + up to 100 imported repos.
+    // todo(ft): look into deduplicating and merging these on the server
+    const suggestedQuery = useSuggestedRepositories({ excludeConfigurations: true });
```
Comment (mustard-mh) [inline components/dashboard/src/data/git-providers/unified-repositories-search-query.ts:40]: It's weired to do this change, I image that `suggestedRespositories` has some logic inside like base on users' activity recommend them with different priority of items.

In current implement, we will filter out configurations and append only via name searching.

Just feel it's weired, not blocking

**Comment by mustard-mh** [comment]
> Workspaces worked well
> 
> * By selecting an SCM repo https://mustardmh-springpetclin-0xi06n0lptu.ws-dev.ft-improve53bed5c2a8.preview.gitpod-dev.com/
> * By selecting an imported repo https://tes-tenk100-va8u05g3lsd.ws-dev.ft-improve53bed5c2a8.preview.gitpod-dev.com/
> * By directly going to the page with an imported repo's URL https://tes-2krepos1000-vyclwfugewd.ws-dev.ft-improve53bed5c2a8.preview.gitpod-dev.com/

@filiptronicek do you expect the third on could choose a configuration to use? I found it's not using a configuration in ListWorkspace

File: components/dashboard/src/data/git-providers/unified-repositories-search-query.ts (hunk 6 above)
Comment (mustard-mh) [reply components/dashboard/src/data/git-providers/unified-repositories-search-query.ts:40]: It leads a condition: I use `bitbucket.gitpod-dev.com/scm/tes/ten-k-1304` to start a workspace, and back to dashboard, the first listed context url in new workspace page is not `1304`

Comment (filiptronicek) [reply components/dashboard/src/data/git-providers/unified-repositories-search-query.ts:40]: That's fair, we should not make this reorder all projects all of a sudden. Reverting this change

File: components/gitpod-db/src/typeorm/project-db-impl.ts (hunk 5 above)
Comment (filiptronicek) [reply components/gitpod-db/src/typeorm/project-db-impl.ts:81]: Sounds good to impose a limit here. Added that in [`1c5a88d` (#20151)](https://github.com/gitpod-io/gitpod/pull/20151/commits/1c5a88db3cf967a5af8feabc9bfd3bec63a0aa2e)

File: components/dashboard/src/data/git-providers/unified-repositories-search-query.ts (hunk 6 above)
Comment (filiptronicek) [reply components/dashboard/src/data/git-providers/unified-repositories-search-query.ts]: [`bbd8559` (#20151)](https://github.com/gitpod-io/gitpod/pull/20151/commits/bbd8559beae585c89f180dd80025a04d65133d75)

**Review by mustard-mh — APPROVED** [review]

File: components/dashboard/src/components/RepositoryFinder.tsx [hunk 7]
```diff
@@ -126,6 +128,10 @@ export default function RepositoryFinder({
                 return repo.configurationId === selectedConfigurationId;
             }
 
+            if (repo.url.endsWith(".git")) {
```
[the remark this answers is not in the document]
Comment (filiptronicek) [reply components/dashboard/src/components/RepositoryFinder.tsx:187]: yes, this piece is duplicated in so many places. We should try to have this somewhere more centralized.

File: components/server/src/api/scm-service-api.ts [hunk 8]
```diff
@@ -80,16 +80,19 @@ export class ScmServiceAPI implements ServiceImpl<typeof ScmServiceInterface> {
...
-            ? this.projectService.getProjects(userId, organizationId)
+            ? this.projectService.getProjects(userId, organizationId, { limit: request.pagination?.pageSize })
             : Promise.resolve([]);
         const workspacesPromise = this.workspaceService.getWorkspaces(userId, { organizationId });
         const repos = await this.scmService.listSuggestedRepositories(userId, { projectsPromise, workspacesPromise });
         return new ListSuggestedRepositoriesResponse({
             repositories: repos.map((r) => this.apiConverter.toSuggestedRepository(r)),
             pagination: new PaginationResponse({
                 nextToken: "",
-                total: repos.length,
```
Comment (filiptronicek) [inline components/server/src/api/scm-service-api.ts:92]: we no longer know

**Pull request merged**